edx_prefectutils/amplitude.py

```diff
@@ -0,0 +1,347 @@
+                requests.exceptions.Timeout,  # This includes read timeout.
+                # In other data pipelines, retrying on read timeout could be dangerous because the events from the
+                # failed request my have been successfully ingested, so retrying can result in duplicate events.
+                # However, we do not mind retrying on read timeout with Amplitude loading because Amplitude will de-dupe
+                # the events anyway.
+            ),
+        )
+        def send_batch(amplitude_events_batch):
+            """
+            Actually make a Batch API request to send a batch of events.
+
+            This is meant to be executed by a worker thread only.  Since we only want to allow a finite number of worker
+            threads to be running API calls simultaneously, the API call is surrounded by logic to acquire/release a
+            semaphore lock.  It's important that this semaphore logic exists *inside* the retry loop because we don't want
+            to block any other workers while this one is in the middle of backing off.
+
+            Args:
+                amplitude_events_batch (iterable of events): 1000 or fewer events formatted for consumption by Amplitude.
+            """
+            available_senders.acquire()
+            r = requests.post(
+                "https://api2.amplitude.com/2/httpapi",
+                json={
+                    "api_key": amplitude_api_key,
+                    "events": amplitude_events_batch,
+                },
+                headers=AMPLITUDE_REQUEST_HEADERS,
+            )
+            r.raise_for_status()
+            available_senders.release()
+
+        def load_device_segment(device_segment):
+            """
+            Load all the events for a given device within the current processing
+            segment.
+
+            Args:
+              segment(iterable of iterable of events):
+                All events for single user or device, batched by 1000.
+            """
+            for api_batch in device_segment:
+                send_batch(api_batch)
+
+        # Launch a thread pool of 128 (twice that of the semaphore limit) so that
+        # we have a few extra threads to pick up the slack whenever a thread gets
+        # stuck backing off due to EPDS rate limiting.  In theory, any time we have
+        # fewer than the maximum number of blocking network calls, we are running
+        # slower than capacity.
+        with ThreadPoolExecutor(max_workers=128) as executor:
+            executor.map(load_device_segment, segmented_by_device_user_and_batched)
+
+        # Update the high watermark.
+        # We do not need to wrap this in retry logic, since that is already
+        # implemented in the prefect source:
+        # https://github.com/PrefectHQ/prefect/blob/da07d255/src/prefect/client/client.py#L755-L756
+        prefect.backend.set_key_value(
+            key="amplitude_high_watermark",
+            value=new_high_watermark
```

Comment:
So when you come to update this for Snowflake, I would write this new_high_watermark to the log file, and not add one to it.   I think that will be more consistent.